pkg/clustermesh/common/namespacewatcher/watcher.go

```diff
@@ -0,0 +1,480 @@
+	allNamespaces := sets.New[string]()
+
+	// Collect namespaces from all processors
+	for _, processor := range nw.processors {
+		namespaces := processor.GetAllNamespaces()
+		for _, ns := range namespaces {
+			allNamespaces.Insert(ns)
+		}
+	}
+
+	return allNamespaces.UnsortedList()
+}
+
+// isNamespaceAnnotated checks if a namespace has the global annotation by querying the resource store
+func (nw *namespaceWatcher) isNamespaceAnnotated(namespace string) bool {
+	ctx := context.Background()
+	store, err := nw.namespaceResource.Store(ctx)
+	if err != nil {
+		return false
+	}
+
+	ns, exists, err := store.GetByKey(resource.Key{Name: namespace})
+	if err != nil || !exists {
+		return false
+	}
+
+	_, hasAnnotation := annotation.Get(ns, annotation.GlobalNamespace)
+	return hasAnnotation
+}
+
+// isNamespaceGlobalByAnnotation checks if a namespace is marked as global by its annotation
+func (nw *namespaceWatcher) isNamespaceGlobalByAnnotation(namespace string) bool {
+	ctx := context.Background()
+	store, err := nw.namespaceResource.Store(ctx)
+	if err != nil {
+		return false
+	}
+
+	ns, exists, err := store.GetByKey(resource.Key{Name: namespace})
+	if err != nil || !exists {
+		return false
+	}
+
+	annotationValue, hasAnnotation := annotation.Get(ns, annotation.GlobalNamespace)
+	if !hasAnnotation {
+		return false
+	}
+
+	return annotationValue == "true"
+}
+
+// isFilteringActive checks if namespace-based filtering is currently active by checking if any namespace has the annotation
+func (nw *namespaceWatcher) isFilteringActive() bool {
+	ctx := context.Background()
+	store, err := nw.namespaceResource.Store(ctx)
+	if err != nil {
+		return false
+	}
+
+	// List all namespaces and check if any have the annotation
+	allNamespaces := store.List()
+	for _, ns := range allNamespaces {
+		if _, hasAnnotation := annotation.Get(ns, annotation.GlobalNamespace); hasAnnotation {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (nw *namespaceWatcher) IsGlobalNamespace(namespace string) bool {
+	nw.mu.RLock()
+	defer nw.mu.RUnlock()
+
+	// If namespace-based filtering is not active (no annotated namespaces),
+	// all namespaces are global for backwards compatibility
+	if !nw.isFilteringActive() {
+		return true
+	}
+
+	// If the namespace is explicitly annotated, use the annotation value
+	if nw.isNamespaceAnnotated(namespace) {
+		return nw.isNamespaceGlobalByAnnotation(namespace)
+	}
+
+	// For non-annotated namespaces when filtering is active, use the default behavior
+	return nw.config.DefaultGlobalNamespace
+}
+
+func (nw *namespaceWatcher) IsFilteringActive() bool {
+	nw.mu.RLock()
+	defer nw.mu.RUnlock()
+	return nw.isFilteringActive()
+}
+
+func (nw *namespaceWatcher) GetGlobalNamespaces() sets.Set[string] {
+	nw.mu.RLock()
+	defer nw.mu.RUnlock()
+
+	// If namespace-based filtering is not active, return empty set to indicate all namespaces are global
+	if !nw.isFilteringActive() {
+		return sets.New[string]()
+	}
```

Comment:
How can a caller distinguish between all namespaces are global and no namespaces are global?